File: src/commands/assets.ts
// Typed Tauri command wrappers for asset operations.

import { invoke } from '@tauri-apps/api/core';
import type {
  AssetDetail,
  AssetSearchFilters,
  AssetSearchResult,
  AssetSort,
} from '../types/asset';

export async function searchAssets(
  filters: AssetSearchFilters = {},
  sort?: AssetSort,
  page = 1,
  pageSize = 50,
): Promise<AssetSearchResult> {
  return invoke<AssetSearchResult>('asset_search', { filters, sort, page, pageSize });
}

export async function getAsset(assetId: string): Promise<AssetDetail> {
  return invoke<AssetDetail>('asset_get', { assetId });
}

export async function deleteAsset(assetId: string): Promise<void> {
  return invoke<void>('asset_delete', { assetId });
}

export async function openAsset(assetId: string, locationId: string): Promise<void> {
  return invoke<void>('asset_open', { assetId, locationId });
}

export async function revealAsset(assetId: string, locationId: string): Promise<void> {
  return invoke<void>('asset_reveal', { assetId, locationId });
}

File: src-tauri/src/commands/assets.rs
//! Tauri command handlers for asset operations.
//! Uses db_read (read-only connection) to avoid blocking the indexer.

use crate::AppState;
use crate::models::asset::{AssetDetail, AssetSearchFilters, AssetSearchResult, AssetSort};

#[tauri::command]
pub async fn asset_search(
    state: tauri::State<'_, AppState>,
    filters: AssetSearchFilters,
    sort: Option<AssetSort>,
    page: Option<i64>,
    page_size: Option<i64>,
) -> Result<AssetSearchResult, String> {
    let conn = state.db_read.lock().map_err(|e| e.to_string())?;
    crate::assets::search_assets(
        &conn,
        &filters,
        sort.as_ref(),
        page.unwrap_or(1),
        page_size.unwrap_or(50),
    ).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn asset_get(
    state: tauri::State<'_, AppState>,
    asset_id: String,
) -> Result<AssetDetail, String> {
    let conn = state.db_read.lock().map_err(|e| e.to_string())?;
    crate::assets::get_asset(&conn, &asset_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn asset_delete(
    state: tauri::State<'_, AppState>,
    asset_id: String,
) -> Result<(), String> {
    let conn = state.db.lock().map_err(|e| e.to_string())?;
    crate::assets::delete_asset(&conn, &asset_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn asset_open(
    state: tauri::State<'_, AppState>,
    asset_id: String,
    location_id: String,
) -> Result<(), String> {
    let file_path: String = {
        let conn = state.db_read.lock().map_err(|e| e.to_string())?;
        conn.query_row(
            "SELECT file_path FROM locations WHERE id = ?1 AND asset_id = ?2",
            rusqlite::params![location_id, asset_id],
            |row| row.get(0),
        ).map_err(|_| "Location not found".to_string())?
    };

    crate::drives::platform::open_with_default_app(std::path::Path::new(&file_path))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn asset_reveal(
    state: tauri::State<'_, AppState>,
    asset_id: String,
    location_id: String,
) -> Result<(), String> {
    let file_path: String = {
        let conn = state.db_read.lock().map_err(|e| e.to_string())?;
        conn.query_row(
            "SELECT file_path FROM locations WHERE id = ?1 AND asset_id = ?2",
            rusqlite::params![location_id, asset_id],
            |row| row.get(0),
        ).map_err(|_| "Location not found".to_string())?
    };

    crate::drives::platform::reveal_in_file_manager(std::path::Path::new(&file_path))
        .map_err(|e| e.to_string())
}
